### ISSUE TITLE
Usage without a strategy

### ISSUE THREAD

**Issue by nouellette** [hapijs/bell#177]
I'm trying to use bell v2 without using a strategy because I want to dynamically change the location setting on the fly depending on an override setting. But I am running into an issues using it.

The problem is at the end of the cycle of exchanging the token for the data. I think it relates to this line of code:
https://github.com/hapijs/bell/blob/master/lib/oauth.js#L229

Because of the async nature of it my code below will just return before the response comes back with the data for the profile. This type of code would work using the strategy use case because the line below would prevent it from moving forward because of the way the response lifecycle is setup.

https://github.com/hapijs/bell/blob/master/lib/oauth.js#L292

Please let me know if I'm doing something wrong by using the plugin in this manner or if there is another clever way of dealing with this situation. Also if you have any kind of examples, that would be helpful. Thank you for your time.

Here is the code:

```
{
      method: 'GET',
      path: '/auth/oauth2callback',
      config: {
        description: 'OAuth2 callback to authenticate & get profile data.'
      },
      handler: function(request, reply){
        var strategy = config.googleStrategy;

        if(request.query && request.query.override_location) {
          var redirectHost = request.query.override_location;
          delete request.query.override_location;
          strategy.location = redirectHost;
        }

        server.plugins.bell.oauth.v2(strategy)(request, reply);

        console.log(request.auth);
      }
}
```

**Comment by ldesplat** [comment]
This is not application to your use case but FYI you can use a strategy to add dynamic parameters
```
allowRuntimeProviderParams - allows passing query parameters from a bell protected endpoint to the auth request. It will merge the query params you pass along with the providerParams and any other predefined ones. Be aware that this will override predefined query parameters! Default to false.
```
---
Now, I haven't gotten to the specifics of this in a few months so I may be saying something stupid.

I think you're almost there, the problem is that you're overriding the location perfectly on the first go around when it does a redirect to the provider (I guess google here). The problem is when we do a POST to get the actual token, then you are not setting the location. I would set your own cookie (in the if statement) which includes redirectHost. Then you can have another if statement (right after this one) that check if (request.query.code is present). If it is, set the strategy.location to your cookie's value and unstate the cookie.

I have not tried it but something like this, should work:
```
    handler: function(request, reply){
        var strategy = config.googleStrategy;

        if(request.query && request.query.override_location) {
          var redirectHost = request.query.override_location;
          delete request.query.override_location;
          strategy.location = redirectHost;
          reply.state('redirectHost', redirectHost);
        }

        if(request.query && request.query.code) {
            strategy.location = request.state.redirectHost;
            reply.unstate('redirectHost');
        }

        server.plugins.bell.oauth.v2(strategy)(request, reply);

        console.log(request.auth);
      }
```

**Comment by nouellette** [comment]
ldesplat, thank you for responding so fast.

allowRuntimeProviderParams didn't allow me to override redirect_uri during my testing. I was able to add additional params but I wasn't able to override the required ones needed by google, I think this is due to the way the merge code for queries works.

https://github.com/hapijs/bell/blob/master/lib/oauth.js#L166

You may be right about needing the location for the second call but I think my main struggle is the async call piece that is preventing me from figuring this out. 

I did some poking around the internet to see if I could find a working example with this kind of approach and I just don't think it's possible with the async calls unless there is a way to simulate the same structure of how the request pipeline works where reply.continue() comes into play. Are you able to create a working example that works with a similar setup that I am using (as shown above)?

**Comment by nouellette** [comment]
@ldesplat I tried your example above and it did allow me not to have to do another redirect to push things forward in the workflow of oauth2 but I'm still facing this issue with the post request to google's token endpoint to get the profile. The hapi endpoint (`/auth/oauth2callback`) handler will return the request before getting the data back from Google. If there was a way to return the promise from the `server.plugins.bell.oauth.v2(strategy)(request, reply);` I think that would solve my use case. But if you see an easier way to achieve this I'm all ears.

**Comment by ldesplat** [comment]
Ok, I understand `reply.continue()` is not meant to be used in an handler method. If it is used in one it just returns 200 and ok so you get to the situation where you think it returns before getting the data back but really, it did get the data but returned nothing back you.

So, I think you can solve this simply by creating your own auth scheme & strategy. Something like:
```
server.auth.scheme('dynamicBell', (server, options) => {
    return {
        authenticate: (request, reply) => {
            // your current handler code here
         }
    };
});
```
Then create your own strategy and pass the options you want that will be passed as options to the scheme. So probably your googleStrategy. Then use your new strategy as you would use bell normally.

I think that should work?

**Comment by nouellette** [comment]
I think I found a better solution that didn't require a lot of changing around the code.

Essentially what I do was kept the call to: 
`server.plugins.bell.oauth.v2(strategy)(request, reply);` 
the only difference is I added a callback to it to look like this: 
```
server.plugins.bell.oauth.v2(strategy)(request, reply, function(profile){
    console.log(profile.credentials);
});
```

So I had to include that `profileDataCallback` here:
https://github.com/hapijs/bell/blob/master/lib/oauth.js#L23

Then I changed the way the reply.continue() was being handled here:
https://github.com/hapijs/bell/blob/master/lib/oauth.js#L109
https://github.com/hapijs/bell/blob/master/lib/oauth.js#L132

I just did a simple check:
if(profileDataCallback){
    return profileDataCallback({ credentials: credentials });
} else {
    reply.continue({ credentials: credentials });
}

Those were the only changes needed inside the plugin, maybe there is a more elegant way of doing this but worked for me without spending to much time tracking down bugs.

Let me know if this is something you would like me to put in a pull request.
